File: src/main/java/com/yogpc/qp/tile/TileBreaker.java
/*
 * Copyright (C) 2012,2013 yogpstop This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 */

package com.yogpc.qp.tile;

import java.util.HashMap;
import java.util.Map;

import cofh.api.tileentity.IInventoryConnection;
import com.yogpc.qp.QuarryPlus;
import javax.annotation.Nullable;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.fml.common.Optional;
import net.minecraftforge.items.CapabilityItemHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.wrapper.InvWrapper;

@Optional.Interface(iface = "cofh.api.inventory.IInventoryConnection", modid = QuarryPlus.Optionals.COFH_modID)
public class TileBreaker extends TileEntity implements IEnchantableTile, IInventory, IInventoryConnection {
    IItemHandler handler = new InvWrapper(this);
    public boolean silktouch = false;
    public byte fortune = 0;

    @Override
    public void readFromNBT(final NBTTagCompound nbttc) {
        super.readFromNBT(nbttc);
        this.silktouch = nbttc.getBoolean("silktouch");
        this.fortune = nbttc.getByte("fortune");
    }

    @Override
    public NBTTagCompound writeToNBT(final NBTTagCompound nbttc) {
        nbttc.setBoolean("silktouch", this.silktouch);
        nbttc.setByte("fortune", this.fortune);
        return super.writeToNBT(nbttc);
    }

    @Override
    public Map<Integer, Byte> getEnchantments() {
        final Map<Integer, Byte> ret = new HashMap<>();
        if (this.fortune > 0)
            ret.put(FortuneID, this.fortune);
        if (this.silktouch)
            ret.put(SilktouchID, (byte) 1);
        return ret;
    }

    @Override
    public void setEnchantent(final short id, final short val) {
        if (id == FortuneID)
            this.fortune = (byte) val;
        else if (id == SilktouchID && val > 0)
            this.silktouch = true;
    }

    @Override
    public void G_reinit() {
    }

    @Override
    public int getSizeInventory() {
        return 1;
    }

    @Override
    public boolean isEmpty() {
        return true;
    }

    @Override
    public ItemStack getStackInSlot(final int i) {// NOTE better way?
        return ItemStack.EMPTY;
    }

    @Override
    public ItemStack decrStackSize(final int i, final int a) {
        return ItemStack.EMPTY;
    }

    @Override
    public ItemStack removeStackFromSlot(int index) {
        return ItemStack.EMPTY;
    }

    @Override
    public void setInventorySlotContents(final int index, final ItemStack stack) {
    }

    @Override
    public int getInventoryStackLimit() {
        return 0;
    }

    @Override
    public boolean isUsableByPlayer(EntityPlayer player) {
        return false;
    }

    @Override
    public void openInventory(EntityPlayer player) {
    }

    @Override
    public void closeInventory(EntityPlayer player) {
    }

    @Override
    public boolean isItemValidForSlot(final int s, final ItemStack stack) {
        return false;
    }

    @Override
    public int getField(int id) {
        return 0;
    }

    @Override
    public void setField(int id, int value) {
    }

    @Override
    public int getFieldCount() {
        return 0;
    }

    @Override
    public void clear() {
    }

    @Override
    @Optional.Method(modid = "CoFHAPI|inventory")
    public ConnectionType canConnectInventory(final EnumFacing arg0) {
        return ConnectionType.FORCE;
    }

    @Override
    public String getName() {
        return "tile.breakerplus.name";
    }

    @Override
    public boolean hasCustomName() {
        return false;
    }

    @Override
    public boolean hasCapability(Capability<?> capability, @Nullable EnumFacing facing) {
        return capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
    }

    @Nullable
    @Override
    public <T> T getCapability(Capability<T> capability, @Nullable EnumFacing facing) {
        if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
            return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast(handler);
        } else {
            return super.getCapability(capability, facing);
        }
    }
}

File: src/main/scala/com/yogpc/qp/tile/TileBreaker.scala
/*
 * Copyright (C) 2012,2013 yogpstop This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 */

package com.yogpc.qp.tile

import java.util

import com.yogpc.qp.gui.TranslationKeys
import com.yogpc.qp.tile.IEnchantableTile._
import javax.annotation.Nullable
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.wrapper.InvWrapper

class TileBreaker extends TileEntity with IEnchantableTile with HasInv {
    private[this] val handler = new InvWrapper(this)
    var silktouch = false
    var fortune: Byte = 0

    override def readFromNBT(nbttc: NBTTagCompound): Unit = {
        super.readFromNBT(nbttc)
        this.silktouch = nbttc.getBoolean("silktouch")
        this.fortune = nbttc.getByte("fortune")
    }

    override def writeToNBT(nbttc: NBTTagCompound): NBTTagCompound = {
        nbttc.setBoolean("silktouch", this.silktouch)
        nbttc.setByte("fortune", this.fortune)
        super.writeToNBT(nbttc)
    }

    override def getEnchantments: util.Map[java.lang.Integer, java.lang.Integer] = {
        val ret = new util.HashMap[java.lang.Integer, java.lang.Integer]
        if (this.fortune > 0) ret.put(Int.box(FortuneID), Int.box(fortune))
        if (this.silktouch) ret.put(Int.box(SilktouchID), Int.box(1))
        ret
    }

    override def setEnchantent(id: Short, value: Short) =
        if (id == FortuneID) this.fortune = value.toByte
        else if (id == SilktouchID) this.silktouch = value > 0

    override def G_reinit() = ()

    override def getName = TranslationKeys.breaker

    override def hasCapability(capability: Capability[_], @Nullable facing: EnumFacing): Boolean =
        (capability eq CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) || super.hasCapability(capability, facing)

    @Nullable override def getCapability[T](capability: Capability[T], @Nullable facing: EnumFacing): T =
        if (capability eq CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast(handler)
        else super.getCapability(capability, facing)
}